### PR TITLE
417 Create a database

### PR DESCRIPTION
**Pull request by bell-pa**
## Context

Trello [417 Create database](https://trello.com/c/kUnIVhHw/417-create-database)

Additional environment variables:
* `DATABASE_ENDPOINT`
* `DATABASE_NAME`
* `DATABASE_USERNAME`
* `DATABASE_PASSWORD`
When the application is deployed to the HMPPS Kubernetes cluster the Helm chart will populate these values from the `dps-rds-instance-output` secret from the namespace.

## Changes in this PR
By commit:
### Refactor
Introduce a `CourseRepository` façade and `InMemoryCourseRepository` implementation
Inject this into `CourseService` and mock the repo in tests using the springmockk library (an extension of Mockk for Spring Boot)

### Refactor
Delegate `Course` and `CourseOffering` lookups to `CourseRepository`, move implementations to `InMemoryCourseRepository`.

### Introduce Spring JPA
Add JPA annotations to `CourseEntity` and add a `CourseEntityRepository` with tests. 
Revise `CourseEntity` to meet JPA requirements.

### Fix logback configuration
\<layout/\> is replaced by \<encoder/\>

### Configure the PostgreSQL JPA dialect
Requires adding PostgreSQL driver for full run-time support.

### Improvement
Drop `strategy = GenerationType.UUID` from `@GeneratedValue` annotations on JPA Entities  because this is the default for the UUID type.

### PostgreSQL integration 
* Add PostgreSQL to docker-compose.yml and align application-local.yml with this. 
* Add PostgreSQL datasource configuration to application.yml. 
* Map Kubnernetes database secrets in Helm values.yaml

### CODE REVIEW

**Review by Gweaton — APPROVED** [review]
This looks good to me, thanks for tidying it up - I found it really easy to review commit by commit.

I'm approving so as not to block you, but please could you fix up that one commit that's slightly out of sync before merging?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/inmemoryrepo/InMemoryCourseRepositoryTest.kt
Comment (bell-pa) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/inmemoryrepo/InMemoryCourseRepositoryTest.kt:13]: In the previous commit `CourseServiceTest` tests an instance of `CourseService`. (The CourseService instance has a CourseRepository instance injected in turn). This demonstrates that the refactoring hasn't broken the CourseService tests. That is, I'm changing code without changing tests.

In the commit above I then re-purpose CourseServiceTest to test InMemoryCourseRepository directly.  So in this commit I'm changing the test, not the code. The CourseService instance is removed from the test and replaced by the repository.  That is why I renamed the test class in this commit because it is testing a different class.

For this reason I think that moving the rename to the previous commit would be wrong.

You could argue that the second commit (the rename commit) should be split in two  - 1) rename and repurpose the tests, 2) Add flyway and flyway migration for CourseEntity.

**Pull request merged**
